### PR TITLE
Feature mirror and amplification

### PR DESCRIPTION
**Pull request by champly**
## mirro support protocal

- Http1
- Http2
- xprotocal [dubbo/bolt]

## amplification support protocal

- Http1
- Http2
- xprotocal [dubbo/bolt]

### CODE REVIEW

**Comment by champly** [comment]
@wangfakang PTAL

**Comment by champly** [comment]
## config

```json
{
	"servers": [
		{
			"default_log_path": "stdout",
			"default_log_level": "DEBUG",
			"listeners": [
				{
					"name": "egress_dubbo",
					"address": "0.0.0.0:20885",
					"bind_port": true,
					"filter_chains": [
						{
							"tls_context": {},
							"filters": [
								{
									"type": "proxy",
									"config": {
										"downstream_protocol": "X",
										"upstream_protocol": "X",
										"router_config_name": "20882",
										"extend_config": {
											"sub_protocol": "dubbo"
										}
									}
								}
							]
						}
					],
					"stream_filters": [
						{
							"type": "mirror",
							"config": {
								"amplification": 10
							}
						}
					]
				},
				{
					"name": "http1",
					"address": "0.0.0.0:8081",
					"bind_port": true,
					"filter_chains": [
						{
							"tls_context": {},
							"filters": [
								{
									"type": "proxy",
									"config": {
										"downstream_protocol": "Http1",
										"upstream_protocol": "Http1",
										"router_config_name": "8080"
									}
								}
							]
						}
					],
					"stream_filters": [
						{
							"type": "mirror",
							"config": {
								"amplification": 10
							}
						}
					]
				},
				{
					"name": "http2",
					"address": "0.0.0.0:8082",
					"bind_port": true,
					"filter_chains": [
						{
							"tls_context": {},
							"filters": [
								{
									"type": "proxy",
									"config": {
										"downstream_protocol": "Http1",
										"upstream_protocol": "Http2",
										"router_config_name": "18080"
									}
								}
							]
						}
					],
					"stream_filters": [
						{
							"type": "mirror",
							"config": {
								"amplification": 20
							}
						}
					]
				}
			],
			"routers": [
				{
					"router_config_name": "20882",
					"virtual_hosts": [
						{
							"name": "provider",
							"domains": [
								"*"
							],
							"routers": [
								{
									"match": {
										"headers": [
											{
												"name": "service",
												"value": ".*"
											}
										]
									},
									"route": {
										"cluster_name": "dubbo_provider_cluster"
									},
									"request_mirror_policies": {
										"cluster": "dubbo_mirror",
										"percent": 30
									}
								}
							]
						}
					]
				},
				{
					"router_config_name": "8080",
					"virtual_hosts": [
						{
							"name": "http1",
							"domains": [
								"*"
							],
							"routers": [
								{
									"match": {
										"prefix": "/"
									},
									"route": {
										"cluster_name": "http1"
									},
									"request_mirror_policies": {
										"cluster": "http1_mirror",
										"percent": 100
									}
								}
							]
						}
					]
				},
				{
					"router_config_name": "18080",
					"virtual_hosts": [
						{
							"name": "http1",
							"domains": [
								"*"
							],
							"routers": [
								{
									"match": {
										"prefix": "/"
									},
									"route": {
										"cluster_name": "http2"
									},
									"request_mirror_policies": {
										"cluster": "http2_mirror",
										"percent": 100
									}
								}
							]
						}
					]
				}
			]
		}
	],
	"cluster_manager": {
		"clusters": [
			{
				"Name": "dubbo_provider_cluster",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes": 32768,
				"hosts": [
					{
						"address": "127.0.0.1:20880"
					}
				]
			},
			{
				"Name": "dubbo_mirror",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes": 32768,
				"hosts": [
					{
						"address": "127.0.0.1:20883"
					}
				]
			},
			{
				"Name": "http1",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes": 32768,
				"hosts": [
					{
						"address": "127.0.0.1:8080"
					}
				]
			},
			{
				"Name": "http1_mirror",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes": 32768,
				"hosts": [
					{
						"address": "127.0.0.1:9090"
					}
				]
			},
			{
				"Name": "http2",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes": 32768,
				"hosts": [
					{
						"address": "127.0.0.1:18080"
					}
				]
			},
			{
				"Name": "http2_mirror",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes": 32768,
				"hosts": [
					{
						"address": "127.0.0.1:19090"
					}
				]
			}
		]
	},
	"admin": {
		"address": {
			"socket_address": {
				"address": "0.0.0.0",
				"port_value": 34902
			}
		}
	},
	"metrics": {
		"shm_zone": "metrics",
		"sinks": [
			{
				"type": "prometheus",
				"config": {
					"port": 15090
				}
			}
		]
	}
}
```

## result

### dubbo

- demo: examples/codes/dubbo/dubbo-examples
- request: localhost:20880
- mirror: localhost:20883

dubbo

### http1

- demo: examples/codes/filter/mirror/http1
- request: localhost:8080
- mirror: localhost:9090

```shell
for i in `seq 1 1000000`;do echo echo `date '+%Y-%m-%d %H:%M:%S'`; curl -v http://localhost:8081 --form 'demo=@/Users/champly/Downloads/ca.crt'; sleep 1;done
```

http

### http2

- demo: examples/codes/filter/mirror/http2
- request:  localhost:18080
- mirror: localhost:19090

**use http1 trans http2**

```shell
for i in `seq 1 1000000`;do echo echo `date '+%Y-%m-%d %H:%M:%S'`; curl -v http://localhost:8082 --form 'demo=@/Users/champly/Downloads/ca.crt'; sleep 1;done
```

http2

File: pkg/xds/conv/convertxds.go [hunk 1]
```diff
@@ -522,7 +522,7 @@ func convertIstioPercentage(percent *xdstype.FractionalPercent) uint32 {
 	case xdstype.FractionalPercent_MILLION:
 		return percent.Numerator / 10000
 	case xdstype.FractionalPercent_TEN_THOUSAND:
-		return percent.Numerator / 100
+		return percent.Numerator / 10
```
Comment (wangfakang) [inline pkg/xds/conv/convertxds.go:525]: why changed the `100` to `10` ?

Comment (champly) [reply pkg/xds/conv/convertxds.go:525]: because the result is percent，if denominator is thousand, the percent will except 10 not 100

Comment (wangfakang) [reply pkg/xds/conv/convertxds.go]: TEN_THOUSAND

Comment (champly) [reply pkg/xds/conv/convertxds.go]: sorry, i read it wrong

File: pkg/xds/conv/convertxds.go [hunk 2]
```diff
@@ -1336,3 +1337,40 @@ func convertTLS(xdsTLSContext interface{}) v2.TLSConfig {
 	config.Status = true
 	return config
 }
+
+func convertMirrorPolicy(xdsRouteAction *xdsroute.RouteAction) *v2.RequestMirrorPolicy {
+
+	// Deprecated: Do not use.
+	// if xdsRouteAction.GetRequestMirrorPolicy() != nil {
```
Comment (wangfakang) [inline pkg/xds/conv/convertxds.go:1344]: Could delete directly.

**Pull request merged**